Store sale offer events for marketplace sale-item collection.

10thfloor opened the pull request:
There is no way to see all items available for sale in the marketplace. 
This PR is an attempt to create such a list by capturing sale-offer events using the events worker.

This PR also removes the `kitty_items` table.

10thfloor commented on kitty-items-js/src/workers/base-event-handler.ts:
```diff
@@ -18,23 +17,21 @@ abstract class BaseEventHandler {
+
     let blockCursor = await this.blockCursorService.findOrCreateLatestBlockCursor(
-      this.eventName,
+      "SALE_OFFER_EVENTS",
```
Not sure why the cursor is synonymous with the event name. Made it generic.

10thfloor commented on kitty-items-js/src/workers/base-event-handler.ts:
```diff
@@ -55,38 +52,43 @@ abstract class BaseEventHandler {
         ]);
-        eventList = await fcl.decode(getEventsResult);
+
+        console.log(JSON.stringify(getEventsResult, null, 2));
```
```suggestion
```

10thfloor commented:
Upon further consideration, we should not be queryin over all events, because perormance will degrade, the more users are sending events to the system.

10thfloor commented on kitty-items-cadence/contracts/KittyItemsMarket.cdc:
```diff
@@ -195,18 +211,25 @@ pub contract KittyItemsMarket {
 
-            emit CollectionInsertedSaleOffer(saleItemID: id, saleItemCollection: self.owner?.address!)
+            emit CollectionInsertedSaleOffer(
+              account: self.owner?.address!, 
```
```suggestion
```

10thfloor commented on kitty-items-cadence/contracts/KittyItemsMarket.cdc:
```diff
@@ -195,18 +211,25 @@ pub contract KittyItemsMarket {
         // Remove and return a SaleOffer from the collection.
         pub fun remove(saleItemID: UInt64): @SaleOffer {
-            emit CollectionRemovedSaleOffer(saleItemID: saleItemID, saleItemCollection: self.owner?.address!)
+            emit CollectionRemovedSaleOffer(itemID: saleItemID, account: self.owner?.address!)
```
```suggestion
            emit CollectionRemovedSaleOffer(saleItemID: saleItemID, saleItemCollection: self.owner?.address!)
```

10thfloor commented on kitty-items-cadence/transactions/kittyItemsMarket/sell_market_item.cdc:
```diff
@@ -31,6 +37,7 @@ transaction(saleItemID: UInt64, saleItemPrice: UFix64) {
         let offer <- KittyItemsMarket.createSaleOffer (
             sellerItemProvider: self.kittyItemsCollection,
             saleItemID: saleItemID,
+            saleItemType: self.kittyItemsCollection.borrow()!.borrowKittyItem(id: saleItemID)!.typeID,
```
Expose the sale item type to make it available to the emitted SaleOffer event.

10thfloor commented on kitty-items-cadence/contracts/Kibble.cdc:
```diff
@@ -1,4 +1,4 @@
-import FungibleToken from "./FungibleToken.cdc"
+import FungibleToken from 0xee82856bf20e2aa6
```
Emulator FT address.

psiemens replied:
Check out this feature: https://docs.onflow.org/flow-cli/initialize-project/#advanced-format

You can actually specify contract aliases in the `flow.json` file

psiemens commented on kitty-items-js/src/services/market.ts:
```diff
@@ -121,24 +158,42 @@ class MarketService {
+    // TODO
+  };
+
+  // storeEvent = (event) => {
+  //   console.log("Got event: ", event);
+  //   return SaleOfferEvent.transaction(async (tx) => {
+  //     try {
+  //       await SaleOfferEvent.query(tx).insert({
+  //         event: event,
+  //       });
+  //     } catch (e) {
+  //       console.log("Could not save event", e);
+  //     }
+  //   });
+  // };
```
Is this needed?

10thfloor replied:
No. Will remove. Unless we want to have a table of events we can audit or some such.

psiemens commented on kitty-items-js/src/services/market.ts:
```diff
@@ -121,24 +158,42 @@ class MarketService {
   };
+
+  findMostRecentSales = () => {
+    // TODO
```
Does this not working anymore?

```
return SaleOffer.query().orderBy("created_at", "desc").limit(20);
```

10thfloor replied:
Should work. This method was implemented for the old SaleOffer table so I may have erased it (Thought I was going to be able to get by just storing events). Can re-add it.

psiemens commented on kitty-items-js/src/workers/base-event-handler.ts:
```diff
@@ -55,38 +52,39 @@ abstract class BaseEventHandler {
+          ),
+        ]);
+
+        eventList = await Promise.all([
+          ...getEventsResult.map((result) => fcl.decode(result)),
         ]);
```
```suggestion
        eventList = await Promise.all(getEventsResult.map((result) => fcl.decode(result)));
```

This should work, right?

10thfloor replied:
I think so, I'll double check, could be a 2d array somehow.

psiemens reviewed (commented):
This is really good! Let's find time tomorrow to go through it in more detail.

10thfloor commented on kitty-items-js/src/workers/base-event-handler.ts:
```diff
@@ -55,38 +52,39 @@ abstract class BaseEventHandler {
-        const blockHeight = getEventsResult.events[i].blockHeight;
-        await this.processAndCatchEvent(blockHeight, eventList, i);
-      }
+      // Note: Events may be "processed" out of order.
+      await Promise.all([
+        ...eventList.map((event) => this.processAndCatchEvent({}, event)),
+      ]);
```
Same comment as above.

contra commented on README.md:
```diff
@@ -20,46 +20,58 @@ deployed on [IPFS](https://docs.ipfs.io/concepts/case-study-fleek/) and the Flow
+### 1. The Flow Emulator
 
-Flow is a next-generation blockchain platform that is designed to be fast and easy to use. If you're new to Flow, consider reading more about it in the [Flow Primer](https://www.onflow.org/primer), or visit the [developer documentation](https://docs.onflow.org/).
+Flow is a next-generation blockchain platform that is designed to be fast and easy to use. If you're new to Flow, consider reading more about it in the [Flow Primer](https://www.onflow.org/primer), or visit the [developer documentation](https://docs.onflow.org/). The emulator simulates Flow's operatio in a singel process you can run locally while developing. Read about the emulator [here](https:/ ...
```
Typo - should be `The emulator simulates Flow's operation in a single process`

10thfloor replied:
Nice catch thank you.

10thfloor replied:
Nice catch thank you.

psiemens commented on api/setup-minter.sh:
```diff
@@ -2,35 +2,35 @@
+#     "recipient": "'${MINTER_ADDRESS}'",
+#     "typeId": 1
+#   }'
 
```
Can you revert back to the old one? It should work as long as the `FLOW_ADDRESS` env var has been set

psiemens commented on api/src/config.ts:
```diff
@@ -6,7 +6,7 @@ const flowAccountErrorMessaage = `
 No Flow account configured.
 
-Did you set FLOW_ADDRESS and FLOW_PRIVATE_KEY?
+Did you export FLOW_ADDRESS and FLOW_PRIVATE_KEY?
```
👍

psiemens commented on api/src/migrations/20210315122922_remove_sale_offer_event_table.ts:
```diff
@@ -0,0 +1,10 @@
+
+
+export async function down(knex: Knex): Promise<void> {
+}
```
The migrations seem a bit off. Are all of these needed? This one's empty for example

In short, what needs to be migrated, and can that migration be done in a single migration file?

psiemens commented on api/src/models/sale-offer.ts:
```diff
@@ -1,29 +1,15 @@
+  saleItemType!: number;
+  saleItemCollection!: string;
+  transactionId!: string;
   price!: number;
```
No more seller address?

10thfloor replied:
`saleItemCollection` is the seller address. Named the same as the variable in the contract.

psiemens commented on web/src/parts/account-item-cluster.comp.js:
```diff
@@ -16,6 +16,8 @@ import {
 } from "@chakra-ui/react"
 
 export const ItemImage = ({type}) => {
+  // Lazy load svg iimages of the kitty items.
```
```suggestion
  // Lazy load SVG images for the kitty items.
```

psiemens commented on web/src/hooks/use-market-items.hook.js:
```diff
@@ -1,36 +1,47 @@
+  const [status, setStatus] = useRecoilState($status(IDLE))
+  const [items, setItems] = useRecoilState($state([]))
+
+  useSWR(url, fetcher, {
+    initialData: items,
+    refreshInterval: 10,
+    errorRetryCount: 10,
+    onLoadingSlow: () => {
+      setStatus(LOADING)
+    },
+    onSuccess: ({latestSaleOffers}) => {
+      setStatus(IDLE)
+      setItems(latestSaleOffers)
+    },
+    onError: error => {
+      console.log("Failed to fetch market items.", error)
+    },
+  })
```
Maybe move this into a separate file for API functions?

I see that this line is defined above, and would probably need to be copied through the project if we start making more API calls:

```js
const fetcher = url => fetch(url).then(res => res.json())
```

psiemens commented on package.json:
```diff
@@ -0,0 +1,24 @@
+  },
+  "scripts": {
+    "start:api": "cd ./api && npm run start:dev",
+    "start:worker": "cd ./api && npm run workers:dev",
```
```suggestion
```

No longer needed, right?

psiemens commented on package.json:
```diff
@@ -0,0 +1,24 @@
+    "start:api": "cd ./api && npm run start:dev",
+    "start:worker": "cd ./api && npm run workers:dev",
+    "start:web": "cd ./web && npm run start:dev",
+    "start": "concurrently \"npm:start:*\""
```
Nice! Does this work?

10thfloor replied:
It does!

psiemens commented on api/src/workers/base-event-handler.ts:
```diff
@@ -55,38 +52,45 @@ abstract class BaseEventHandler {
-          getEvents(this.eventName, fromBlock, toBlock),
-        ]);
-        eventList = await fcl.decode(getEventsResult);
+        getEventsResult = await Promise.all([
```
I'm still confused as to why `Promise.all([...list])` is needed instead of just `Promise.all(list)`

psiemens reviewed (changes requested):
Looking good! Going to demo this now.

10thfloor commented on .gitignore:
```diff
@@ -1,26 +1,33 @@
+/api/node_modules/
+/web/node_modules/
+
+/.pnp
+.pnp.js
```
.pnp.js obviously ;) No idea actually.

10thfloor commented on .gitignore:
```diff
@@ -1,26 +1,33 @@
+.DS_Store
+
+#.env files
+/**/.env
+/**/.env.local
+/**/.env.development.local
+/**/.env.test.local
+/**/.env.production.local
```
Ah yes, was experimenting and forgot to revert.

10thfloor commented on api/.env.example:
```diff
@@ -5,7 +5,7 @@
 MINTER_ADDRESS=${FLOW_ADDRESS}
 MINTER_PRIVATE_KEY=${FLOW_PRIVATE_KEY}
-MINTER_ACCOUNT_KEY_INDEX=${FLOW_ACCOUNT_KEY_INDEX}
+MINTER_ACCOUNT_KEY_INDEX=${FLOW_ADDRESS_KEY_INDEX}
```
I must have changed between account and address a dozen times. Account makes more sense to me as well.

10thfloor commented on api/README.md:
```diff
@@ -65,6 +65,8 @@ _💡 Learn more about `Vault` and `Collection` resources [in this tutorial](htt
 
 Run this script to set up the minter account and mint an initial supply of Kibble and Kitty Items:
 
+*Note: Before you run this script, you'll need to open it and add a value for `FLOW_ADDRESS`*
```
```suggestion
```

10thfloor replied:
```suggestion
```

10thfloor commented on .gitignore:
```diff
@@ -1,26 +1,33 @@
+.DS_Store
+
+#.env files
+/**/.env
+/**/.env.local
+/**/.env.development.local
+/**/.env.test.local
+/**/.env.production.local
```
```suggestion
.env
.env.local
.env.development.local
.env.test.local
.env.production.local
```

contra commented:
Hey @10thfloor - Been following along and incorporating these changes into something we're working on, hope you don't mind an out of left field review.

#### Questions

- Is signing in with the dapper wallet actually working for this project or are you testing using blocto?
- The error management in the worker seems to log errors but continues on to mark the block as processed successfully, is this intended?

#### Issues

- Noticed an issue where the worker works fine, but then after about 24hrs starts throwing errors. Creating a new testnet address from scratch solves it, but repeats again after 24hrs.

```
Error: failed to retrieve events from execution nodes [execution-5f6c73a22445d7d958c6a37c1f3be99c72cacd39894a3e46d6647a9adb007b4d@execution-001.devnet20.nodes.onflow.org:3569=1000 execution-a1f059a3e00347e585658e4c400be1580ba2e2f6d1f39387ad62e33f3f4d246d@execution-002.devnet20.nodes.onflow.org:3569=1000]: 2 errors occurred:
  * rpc error: code = DeadlineExceeded desc = context deadline exceeded
  * rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

- Having an issue where I'm not able to list an item for sale, but this could be specific to our project - curious if you've run into this before: https://flow-view-source.com/testnet/tx/8096284302af7d00c63f29bbe6b7b62bc333a5a94fc9a3921c13494b07601d03

- There seems to be a race condition where react renders the application before the DOM ready, which causes fcl to throw an error. Sometimes refreshing a page that requires you to be logged in will redirect you away as if you were logged out, since the user call via fcl failed. Adding a check via `domready` solved this on our fork.

#### Misc

- knex is pretty out of date on the API, we upgraded it on our fork - let me know if you want me to submit a PR to upstream that
- The kitty-item model is unused on the backend from what I can tell, not sure if this is intended or not
- Some instructions about how to deploy this to mainnet would be great.
- The flow for iterating on contracts in this project is a bit rough - if you change any of the data structures even at a basic level the contracts will all fail in odd ways. We adjusted our contracts so that account initialization wipes out existing data which makes it much quicker to iterate on contracts, the alternative is you have to change all of your storage locations to a new one (like this repo does adding `001` to the end) + make a new testnet address. Maybe something worth considering as a part of this PR or a future one is the ergonomics for iterating on contracts rapidly.

Overall this was a major improvement, makes the project much easier to get up and going and fixes most (or all) of the issues we had testing this out earlier on. Thank you!

10thfloor commented:
@contra Great to have you looking at this. I'll try to provide some answers here: 

> Is signing in with the dapper wallet actually working for this project or are you testing using blocto?

At the moment, TopShot is the only app that uses Dapper, and it's a bespoke wallet service for that app specifically. The Dapper Flow wallet, while planned, is still some time away from being available to the public. We do testing with Blocto. Their wallet works really well and it's the only wallet that supports Testnet at the moment. 

> The error management in the worker seems to log errors but continues on to mark the block as processed successfully, is this intended?

This is intended, but not recommended 😅  Kitty Items isn't meant to be an example of a 'production grade' application so you might find that error handling isn't well designed at the moment. Our aim with the app will be to develop and refine in stages. First, get a working example with all the parts, and then hone details over time, based on suggestions from folks who are poking it, like yourself. The event worker is probably the least refined component of the app at the moment. If you're blocked or have a suggestion on how to improve it we welcome you to create a PR and issues of course.

> knex is pretty out of date on the API, we upgraded it on our fork - let me know if you want me to submit a PR to upstream that

Please do! 🙏

> The kitty-item model is unused on the backend from what I can tell, not sure if this is intended or not

psiemens reviewed (commented):
I cleaned up the SQL migrations a bit -- I think they might be too confusing for people trying to understand this code, and since they tear the entire DB down anyway, I think it's probably just easier for any existing users to wipe their DB before trying the new code.

I also switched from camelCase to snake_case names for SQL fields because that's more conventional (and consistent with the block cursor table).

I think we should rename `CollectionInsertedSaleOffer.saleItemCollection` because I keep getting confused by this. I'm worried users will, too.

Could we just simplify the event entirely?

```js
pub event CollectionInsertedSaleOffer(
  itemID: UInt64, 
  itemType: UInt64, 
  seller: Address, 
  price: UFix64
)
``` 

Also, I'm not seeing the marketplace in the web app UI -- is that WIP?

10thfloor commented on api/src/workers/sale-offer-handler.ts:
```diff
@@ -1,30 +1,36 @@
   ) {
-    super(blockCursorService, flowService, eventName);
+    super(blockCursorService, flowService, eventNames);
+    this.config = getConfig();
```
Yes. This felt awkward when I wrote it. Curious about what side-effects a for the `getConfig` function? 

I think `super()` has to be the first statement in the constructor so this might not work, but coupling the events to this class is more consistent so will try to make that work.

10thfloor replied:
Ok, I updated the class. Let me know what you think.

10thfloor commented:
> I cleaned up the SQL migrations a bit -- I think they might be too confusing for people trying to understand this code, and since they tear the entire DB down anyway, I think it's probably just easier for any existing users to wipe their DB before trying the new code.
> 
> I also switched from camelCase to snake_case names for SQL fields because that's more conventional (and consistent with the block cursor table).
> 
> I think we should rename `CollectionInsertedSaleOffer.saleItemCollection` because I keep getting confused by this. I'm worried users will, too.
> 
> Could we just simplify the event entirely?
> 
> ```js
> pub event CollectionInsertedSaleOffer(
>   itemID: UInt64, 
>   itemType: UInt64, 
>   seller: Address, 
>   price: UFix64
> )
> ```
> 
> Also, I'm not seeing the marketplace in the web app UI -- is that WIP?

The marketplace UI is there, might not be  fetching SaleOffers properly after all our latest updates. I'll debug this. 
Agreed on changing the event details as well.

psiemens commented on api/src/workers/base-event-handler.ts:
```diff
@@ -1,40 +1,38 @@
+
     let blockCursor = await this.blockCursorService.findOrCreateLatestBlockCursor(
-      this.eventName,
+      "SALE_OFFER_EVENTS",
```
This should probably be a parameter, given that this is the base handler

psiemens commented on api/src/workers/base-event-handler.ts:
```diff
@@ -55,38 +53,45 @@ abstract class BaseEventHandler {
-        const blockHeight = getEventsResult.events[i].blockHeight;
-        await this.processAndCatchEvent(blockHeight, eventList, i);
-      }
+      // Note: Events may be "processed" out of order.
```
Maybe this should be done sequentially (without `Promise.all`)? As of now there are no logical dependencies between events, but based on my experience working with Ethereum events for CK, I think that could very likely change in the future. 

I also could see many extensions of this example needing to ensure strict ordering when processing events, and they might not realize that this handler sends them through in parallel.

10thfloor replied:
Yes, i had this on a TODO as well. We can make this sequential.

psiemens commented on api/src/workers/sale-offer-handler.ts:
```diff
@@ -1,30 +1,48 @@
+    flowService: FlowService
   ) {
-    super(blockCursorService, flowService, eventName);
+    super(blockCursorService, flowService, []);
+
+    this.eventCollectionInsertedSaleOffer = `A.${fcl.sansPrefix(
+      marketService.marketAddress
+    )}.KittyItemsMarket.CollectionInsertedSaleOffer`;
+
+    this.eventCollectionRemovedSaleOffer = `A.${fcl.sansPrefix(
+      marketService.marketAddress
+    )}.KittyItemsMarket.CollectionRemovedSaleOffer`;
+
+    this.eventNames = [
+      this.eventCollectionInsertedSaleOffer,
+      this.eventCollectionRemovedSaleOffer,
+    ];
```
I think you can do this as long as `this` isn't accessed/mutated before the call to `super`.

```suggestion
    const eventCollectionInsertedSaleOffer = `A.${fcl.sansPrefix(
      marketService.marketAddress
    )}.KittyItemsMarket.CollectionInsertedSaleOffer`;
    
    const eventCollectionRemovedSaleOffer = `A.${fcl.sansPrefix(
      marketService.marketAddress
    )}.KittyItemsMarket.CollectionRemovedSaleOffer`;
    
    const eventNames = [
      eventCollectionInsertedSaleOffer,
      eventCollectionRemovedSaleOffer,
    ];
    
    super(blockCursorService, flowService, eventNames);

    this.eventCollectionInsertedSaleOffer = eventCollectionInsertedSaleOffer;
    this.eventCollectionRemovedSaleOffer = eventCollectionRemovedSaleOffer;
```

10thfloor replied:
No, does not compile :(

psiemens commented on cadence/contracts/KittyItemsMarket.cdc:
```diff
@@ -54,7 +60,7 @@ pub contract KittyItemsMarket {
     pub resource interface SaleOfferPublicView {
         pub var saleCompleted: Bool
-        pub let saleItemID: UInt64
+        pub let saleItemId: UInt64
```
The rest of these contracts used caps (`ID`) instead of `Id`, and I think that's becoming a de-facto standard in Cadence. I know that `Id` is standard in JS, but I'd recommend keeping this as-is in Cadence.

10thfloor replied:
Ok, wasn't aware this was a Cadence thing. Will revert.

psiemens commented on web/package.json:
```diff
@@ -21,10 +21,11 @@
   },
   "scripts": {
-    "start:dev": "react-scripts start",
+    "start:dev": "PORT=3001 react-scripts start",
```
`PORT` is set in `.env.example` now, so I don't think this is needed, right?

10thfloor replied:
It wasn't picking it up for some reason. I'll double check.

psiemens commented on web/src/hooks/use-market-item.hook.js:
```diff
@@ -3,12 +3,17 @@ import {sansPrefix} from "@onflow/fcl"
 import {buyMarketItem} from "../flow/buy-market-item.tx"
 import {cancelMarketListing} from "../flow/cancel-market-listing.tx"
 
+class SaleOffer {
+  constructor(accountItem) {
+    if(!accountItem)
+  }
+}
```
How does this work? I'm curious about the `if(!accountItem)` line -- is that shorthand?

10thfloor replied:
This is incomplete. 😅

10thfloor replied:
Not going to use a class for this, just going to create a normalizing function. Something that converts from snake case (result from the db query) to camel case.

psiemens commented on web/src/hooks/use-market-item.hook.js:
```diff
@@ -21,7 +26,14 @@ export const $state = atomFamily({
+    get: key => async ({ get }) => {
+      const accountItem = await fetchMarketItem(...expand(key))
+      if (!accountItem) {
+        // set the status for the key to PROCESSING
+        // Because the db and contract are out of sync. 
```
Can this be removed?

10thfloor replied:
Yes, this is WIP.

psiemens commented on web/src/parts/account-item-cluster.comp.js:
```diff
@@ -53,7 +55,7 @@ export function AccountItemCluster({address, id}) {
         <>
           {!item.forSale ? (
-            <Td isNumeric maxW="50px">
+            <Td isNumeric maxW="none">
```
Is "none" the same as just omitting this attribute entirely?

psiemens commented on web/src/parts/account-item-cluster.comp.js:
```diff
@@ -67,12 +69,12 @@ export function AccountItemCluster({address, id}) {
             </Td>
           ) : (
-            <Td isNumeric maxW="50px">
+            <Td isNumeric maxW="none">
```
Is "none" the same as just omitting this attribute entirely?

10thfloor replied:
I think there is a default max width on these components that I need to override. Will double check.

psiemens commented:
Thanks for fixing that field name typo in the DB!

psiemens commented on api/src/services/block-cursor.ts:
```diff
@@ -1,16 +1,14 @@
+  async findOrCreateLatestBlockCursor(latestBlockHeight: number) {
     let blockCursor = await BlockCursor.query().findOne({
-      flow_event_name: eventName,
+      cursor_label: this.cursorLable,
```
```suggestion
      cursor_label: this.cursorLabel,
```

10thfloor commented on api/src/services/market.ts:
```diff
@@ -133,22 +139,34 @@ class MarketService {
-        },
-      ]);
-      return saleOffers[0];
+      const offers = await SaleOffer.query(tx).select("*");
```
Needs pagination. Can leave for now.

10thfloor commented on api/src/workers/base-event-handler.ts:
```diff
@@ -55,39 +50,30 @@ abstract class BaseEventHandler {
+          const result = await send([getEvents(event, fromBlock, toBlock)]);
+          const decoded = await fcl.decode(result);
+          if (decoded.length) {
+            decoded.forEach((event) => this.onEvent(event));
```
```suggestion
            decoded.forEach(async (event) => await this.onEvent(event));
```

10thfloor commented on api/src/workers/sale-offer-handler.ts:
```diff
@@ -1,30 +1,49 @@
+    )}.KittyItemsMarket.CollectionRemovedSaleOffer`;
+
+    this.eventNames = [
+      // Events will be processes in the order they are added here.
```
```suggestion
      // Events will be processed in the order they are added here.
```

10thfloor commented on api/src/workers/sale-offer-handler.ts:
```diff
@@ -1,30 +1,49 @@
+
+    switch (event.type) {
+      case this.eventCollectionInsertedSaleOffer:
+        this.marketService.addSaleOffer(event);
```
```suggestion
        await this.marketService.addSaleOffer(event);
```

10thfloor commented on api/src/workers/sale-offer-handler.ts:
```diff
@@ -1,30 +1,49 @@
+        this.marketService.addSaleOffer(event);
+        break;
+      case this.eventCollectionRemovedSaleOffer:
+        this.marketService.removeSaleOffer(event);
```
```suggestion
        await this.marketService.removeSaleOffer(event);
```

10thfloor commented on cadence/contracts/KittyItemsMarket.cdc:
```diff
@@ -191,27 +206,33 @@ pub contract KittyItemsMarket {
          pub fun insert(offer: @KittyItemsMarket.SaleOffer) {
-            let id: UInt64 = offer.saleItemID
-
+            let id: UInt64 = offer.itemID
+            let price: UFix64 = offer.price
+            let itemType: UInt64 = offer.typeID
             // add the new offer to the dictionary which removes the old one
             let oldOffer <- self.saleOffers[id] <- offer
```
```suggestion
            let itemID: UInt64 = offer.itemID
            let typeID: UInt64 = offer.typeID
            let price: UFix64 = offer.price

            // add the new offer to the dictionary which removes the old one
            let oldOffer <- self.saleOffers[itemID] <- offer
```

10thfloor commented on cadence/contracts/KittyItemsMarket.cdc:
```diff
@@ -191,27 +206,33 @@ pub contract KittyItemsMarket {
 
-            emit CollectionInsertedSaleOffer(saleItemID: id, saleItemCollection: self.owner?.address!)
+            emit CollectionInsertedSaleOffer(
+              itemID: id,
+              typeID: itemType,
+              owner:self.owner?.address!,
+              price: price
```
```suggestion
              itemID: itemID,
              typeID:typeID,
              owner:self.owner?.address!,
              price: price
```

10thfloor commented on web/src/util/normalize-item.js:
```diff
@@ -0,0 +1,12 @@
+      typeID: item.sale_item_type,
+      owner: item.sale_item_owner,
+      price: item.sale_price,
+      txID: item.txID,
```
```suggestion
      txID: item.transaction_id,
```

psiemens commented on api/src/services/block-cursor.ts:
```diff
@@ -1,16 +1,14 @@
-    eventName: string,
-    latestBlockHeight: number
-  ) {
+  private readonly cursorLabel: string = "BLOCK_CURSOR";
```
What's the purpose of the label if there's also a cursor ID? It looks like there was originally supposed to be one cursor per event type, but now there's one global cursor. Is that right?

I would make `cursorLabel` a parameter and allow `BlockCursorService` to maintain multiple cursors. Even if that's the case, I think it's totally okay for the event handler to use the same cursor for all events right now.

10thfloor replied:
I thought about this, but didn;t see the sense of maintaining 2 cursors, since we get all the events we're interested in in a single query?

The label is there, because we don't know the id beforehand, so can't use it in the JS to query the table for the cursor.

psiemens commented on web/src/util/normalize-item.js:
```diff
@@ -0,0 +1,12 @@
+export default function normalizeItem(item) {
+  return Object.assign(
```
Out of curiosity, what's the need for `Object.assign` here?

10thfloor replied:
Ah yes, was merging objects, but decided not to. Can just be a plain object literal.

10thfloor replied:
Object doesn't throw an error on null or undefined sources, so it; just reducing the chance for thrown excpeptions.

psiemens reviewed (commented):
Code looks good. Going to run through the setup one more time before approving 👍

10thfloor commented on web/src/util/normalize-item.js:
```diff
@@ -0,0 +1,12 @@
+export default function normalizeItem(item) {
+  return Object.assign(
+    {},
+    {
+      itemID: item.sale_item_id,
+      typeID: item.sale_item_type,
+      owner: item.sale_item_owner,
+      price: item.sale_price,
+      txID: item.transaction_id,
+    }
+  )
```
```suggestion
  return    {
      itemID: item.sale_item_id,
      typeID: item.sale_item_type,
      owner: item.sale_item_owner,
      price: item.sale_price,
      txID: item.transaction_id,
    }
```

psiemens commented on api/src/workers/base-event-handler.ts:
```diff
@@ -1,129 +1,95 @@
-    if (toBlock > latestBlockHeight.height) {
-      toBlock = latestBlockHeight.height;
-    }
+    let toBlock = latestBlockHeight.height;
```
I think it's probably a good idea to keep the `stepSize` parameter. If somebody stops their server for 1000 blocks and restarts it, this code will try to catch query all 1000 blocks at once. This will fail and the server enters a crash loop.

psiemens commented on api/src/workers/base-event-handler.ts:
```diff
@@ -1,129 +1,95 @@
-        continue;
-      }
-
-      // make sure we query the access node only when we have a substantial gap
-      // between from and to block ranges; currently if we query the access node
-      // when the gap is too narrow, it returns an error.
```
I think this might be happening again. I'm getting this error fairly frequently:


> Error retrieving events for block range fromBlock=26996926 toBlock=26996928 Error: 
start height 26996926 is greater than the last sealed block height 26996925

I think it probably makes sense to bring back `blockThreshold`. However, you can reduce it to something smaller than 100. I think 5 or 10 would be fine.

10thfloor replied:
It will query from the block where it last saw an event it cares about, up to the latest block, so we don't need a threshold. We could set a limit ... say, if the block range is more than 1000, than move the window to 1000 from the latest block.

psiemens reviewed (commented):
Something strange is happening when I test out the FE app.

After listing an item for sale:
- The "List" button changed to "Unlist" but the spinner never stops
- My laptop fans starting going crazy. I tried to switch to the "Marketplace Items" tab -- it was very slow to change tabs, and after a few seconds Chrome disabled the tab
- I then tried reloading the page in Firefox too, but it also froze and the tab was killed by FF

Any chance there's a runaway loop in the frontend code?

psiemens reviewed (approved):
:shipit:
